Fix Flax params dtype

patil-suraj opened the pull request:
# What does this PR do?

The `dtype` argument in flax models is used ambiguously. This argument is actually supposed to specify the `dtype` of computation and not the `dtype` of model parameters. But in some models/modules, it's passed to kernel_initializers  which causes the `kernel` parameters to be initialized with that `dtype`. This causes the following issues

- in flax models, we don't pass `bias_init` to `Dense` layers since the default value is as expected by our models. So if we pass `dtype=jnp.bfloat16` it's only passed to `kernel_init`, so for a dense layer the kernel params are in `bfloat16` while the `bias` params are in `fp32`
-  This also causes issues with saving and loading models as explained in #12534

This PR corrects the usage of `dtype` in flax models and  adds `to_bf16`, `to_fp16` and `to_fp32` methods in `FlaxPreTrainedModel`. These methods could accept any arbitrary params tree and change its `dtype`. So if users want they could keep certain params in bf16 and certain others in fp32 however they like, by just passing the right parameters to these methods.

To allow keeping only certain params in half-precision the  `to_bf16` method accepts a mask that specifies what params to keep in `bf16` and what params in `fp32`. For example

```python3
import jax
import jax.numpy as jnp
from flax.core.frozen_dict import freeze, unfreeze
from flax.traverse_util import flatten_dict, unflatten_dict
from transformers import FlaxBertModel, BertConfig

config = BertConfig(num_hidden_layers=1)
model = FlaxBertModel(config, dtype=jnp.dtype("bfloat16"))

# keep layer norm in fp32
def mask_fn(params):
    flat_params = flatten_dict(params)
    flat_mask = {path: (path[-2:] != ("LayerNorm", "bias") and path[-2:] != ("LayerNorm", "scale")) for path in flat_params}
    return unflatten_dict(flat_mask)

mask = mask_fn(model.params)
params = model.to_bf16(model.params, mask)

jax.eval_shape(lambda x : x, freeze(params)) # view the dtypes
```

- This PR also fixes an issue in some models where the `dtype` was never passed to some modules, so those modules were always doing computation in fp32 even if the user passed `bf16` or `fp16` dtype .

- This should now help enable mixed-precision training in flax models as we can keep the params and computation `dtype` separate. 

---

🚨  **BREAKING CHANGE** 🚨
**Note that: this will be a breaking change since the meaning of `dtype` is now changed and it's only used to specify the data type of computation and does not influence the data type of model parameters.**

patil-suraj committed:
fix inits

patil-suraj committed:
fix embed dtype

patil-suraj committed:
fix embed dtype

patil-suraj committed:
add test to check default dtype

patil-suraj committed:
quality

patil-suraj committed:
add type conversion methods for flax models

sgugger reviewed (approved):
I think I agree with your solution @patil-suraj but I'll defer to @patrickvonplaten 's final say as I don't know Jax/Flax enough to have a strong opinion.

Initializing a model in bfloat16 of float16 seems like the wrong thing to do.

patrickvonplaten commented:
I like the design and I think it follows jax-design quite nicely (similar to how optax optimizers mask certain weights: https://github.com/huggingface/transformers/blob/1fec32adc6a4840123d5ec5ff5cf419c02342b5a/examples/flax/summarization/run_summarization_flax.py#L588)

This PR will necessarly have some breaking changes as after it loading a model with `dtype=bfloat16` won't convert the weights into bfloat16 anymore, so we should announce it well. 

Also it would be great if @avital could maybe quickly give his opinion on the API here

avital commented:
Hi folks, sorry for the radio silence, I'm back now. @jheek has thought carefully about the exact meaning of dtypes in Flax modules so I'd like to hear his take on this confusion.

jheek commented:
I think masking is the right approach here. The right dtype is very context dependent. During inference half precision is almost always better while during training it's pretty much never worth it. And then of course there is fine-tuning where the masked weights are basically in inference mode. The mask based API captures this complexity really well.

patil-suraj committed:
more robust casting

patil-suraj committed:
cast sinusoidal positions

patil-suraj committed:
update pegasus

patil-suraj committed:
update albert

patil-suraj commented:
just noticed a sneaky bug in some flax models, the `dtype` is never passed to some modules, for example here in bart, 
https://github.com/huggingface/transformers/blob/3fbb55c75779824aacfc43067f0892674a9cfbc6/src/transformers/models/bart/modeling_flax_bart.py#L400-L405
attention never receives  `dtype`, so it’s always in `fp32` even if the user passed `bf16` .

same with T5 here
https://github.com/huggingface/transformers/blob/3fbb55c75779824aacfc43067f0892674a9cfbc6/src/transformers/models/t5/modeling_flax_t5.py#L1368

@patrickvonplaten @sgugger I propose we make `dtype` required for all modules except the using facing once? So all main model classes will have a default type (`fp32`) but for all other submodules make it required to avoid such bugs.

patil-suraj committed:
update test

patil-suraj committed:
make sure dtype is passed to every module

patil-suraj committed:
style

patil-suraj committed:
fix electra dense

patil-suraj committed:
fix t5

patil-suraj committed:
quality

patil-suraj committed:
add more tests

patil-suraj committed:
better name

patil-suraj committed:
use the dtype for lm head computation

patrickvonplaten commented on src/transformers/modeling_flax_utils.py:
```diff
@@ -147,6 +147,36 @@ def params(self, params: Union[Dict, FrozenDict]):
             )
         self._params = params
 
+    def _cast_floating_to(self, params: Union[Dict, FrozenDict], dtype: jnp.dtype, mask: Any = None) -> Any:
```
can we add a docstring here?

patrickvonplaten commented on src/transformers/modeling_flax_utils.py:
```diff
@@ -147,6 +147,36 @@ def params(self, params: Union[Dict, FrozenDict]):
+    def _cast_floating_to(self, params: Union[Dict, FrozenDict], dtype: jnp.dtype, mask: Any = None) -> Any:
+
+        # taken from https://github.com/deepmind/jmp/blob/3a8318abc3292be38582794dbf7b094e6583b192/jmp/_src/policy.py#L27
+        def conditional_cast(param):
```
why is that conditional cast needed exactly? Maybe add this in docstring

patrickvonplaten commented on tests/test_modeling_flax_common.py:
```diff
@@ -604,6 +604,74 @@ def test_load_with_mismatched_shapes(self):
                     logits = new_model(**inputs_dict)["logits"]
                     self.assertEqual(logits.shape[1], 42)
 
+    def test_default_params_dtype(self):
```
very nice test!

patrickvonplaten reviewed (approved):
Looks good to me! It would be nice if we could add a couple more tests though :-)

patil-suraj committed:
Merge branch 'master' into fix-flax-dtype

patil-suraj committed:
fix albert

patil-suraj committed:
style

patil-suraj committed:
fix albert embed dtype

patil-suraj committed:
more tests

patil-suraj committed:
fix vision enc-dec

patil-suraj committed:
cleanup

patil-suraj committed:
fix embed dtype pegasus

patil-suraj committed:
fix default param test

patil-suraj committed:
doc

patil-suraj committed:
update template

patil-suraj committed:
fix final_logits_bias dtype

patil-suraj commented:
Hey @patrickvonplaten !

- added a couple more tests as you suggested
- updated the templates
- ran tests on both GPU and TPU and they pass 

Would be awesome if you could take quick final look :)

patrickvonplaten commented on src/transformers/modeling_flax_utils.py:
````diff
@@ -154,6 +154,115 @@ def params(self, params: Union[Dict, FrozenDict]):
 
+    def _cast_floating_to(self, params: Union[Dict, FrozenDict], dtype: jnp.dtype, mask: Any = None) -> Any:
+        """
+        Helper method to cast the floating-point values from given ``params`` tree to the given ```dtype```.
````
```suggestion
        Helper method to cast floating-point values of given parameter (``params``) dictionary to given ```dtype```.
```

patrickvonplaten commented on src/transformers/modeling_flax_utils.py:
```diff
@@ -154,6 +154,115 @@ def params(self, params: Union[Dict, FrozenDict]):
+
+    def to_bf16(self, params: Union[Dict, FrozenDict], mask: Any = None):
+        r"""
+        Cast the floating-point ``parmas`` to ``jax.numpy.bfloat16``. This method can be used to explicitly convert the
```
```suggestion
        Cast the floating-point ``params`` to ``jax.numpy.bfloat16``. This method can be used to explicitly convert the
```

patrickvonplaten commented on src/transformers/modeling_flax_utils.py:
```diff
@@ -154,6 +154,115 @@ def params(self, params: Union[Dict, FrozenDict]):
+    def to_bf16(self, params: Union[Dict, FrozenDict], mask: Any = None):
+        r"""
+        Cast the floating-point ``parmas`` to ``jax.numpy.bfloat16``. This method can be used to explicitly convert the
+        model paramters to ``bfloat16``.
```
```suggestion
        model paramters to bfloat16 precision.
```

patrickvonplaten commented on src/transformers/modeling_flax_utils.py:
```diff
@@ -154,6 +154,115 @@ def params(self, params: Union[Dict, FrozenDict]):
+        Examples::
+
+            >>> from transformers import FlaxBertModel
+            >>> # Download model and configuration from huggingface.co
```
```suggestion
            >>> # load model
```

patrickvonplaten commented on src/transformers/modeling_flax_utils.py:
```diff
@@ -154,6 +154,115 @@ def params(self, params: Union[Dict, FrozenDict]):
+            >>> from transformers import FlaxBertModel
+            >>> # Download model and configuration from huggingface.co
+            >>> model = FlaxBertModel.from_pretrained('bert-base-cased')
+            >>> # By default, the model params will be in fp32, to cast these to bfloat16
```
```suggestion
            >>> # By default, the model parameters will be in fp32 precision, to cast these to bfloat16 precision
```

patrickvonplaten commented on src/transformers/modeling_flax_utils.py:
```diff
@@ -154,6 +154,115 @@ def params(self, params: Union[Dict, FrozenDict]):
+            >>> # If you want don't want to cast certain parameters (for example layer norm bias and scale)
+            >>> # then pass the mask as follows
+            >>> from flax import traverse_util
+            >>> model = FlaxBertModel.from_pretrained('bert-base-cased')
```
```suggestion
```

patrickvonplaten commented on src/transformers/modeling_flax_utils.py:
```diff
@@ -154,6 +154,115 @@ def params(self, params: Union[Dict, FrozenDict]):
+            >>> model = FlaxBertModel.from_pretrained('bert-base-cased')
+            >>> # By default, the model params will be in fp32, to cast these to bfloat16
+            >>> params = model.to_bf16(model.params)
+            >>> # If you want don't want to cast certain parameters (for example layer norm bias and scale)
```
think the example below leaves all biases in fp32 not just layer norm bias no?

patrickvonplaten commented on src/transformers/modeling_flax_utils.py:
```diff
@@ -154,6 +154,115 @@ def params(self, params: Union[Dict, FrozenDict]):
+            >>> # Download model and configuration from huggingface.co
+            >>> model = FlaxBertModel.from_pretrained('bert-base-cased')
+            >>> # By default, the model params will be in fp32, to cast these to bfloat16
+            >>> params = model.to_bf16(model.params)
```
```suggestion
            >>> model.params = model.to_bf16(model.params)
```

Would it maybe make more sense to directly set the parameters equal to the model's parameters? This way it's clear that the model parameters have to be set to the function's output....If I understand correctly `to_bfloat16(...)` doesn't change the input parameters in place, but creates a new dict no?

patrickvonplaten commented on src/transformers/modeling_flax_utils.py:
```diff
@@ -154,6 +154,115 @@ def params(self, params: Union[Dict, FrozenDict]):
+            >>> flat_params = traverse_util.flatten_dict(model.params)
+            >>> mask = {path: (path[-1] != "bias" and path[-2:] != ("LayerNorm", "scale")) for path in flat_params}
+            >>> mask = traverse_util.unflatten_dict(mask)
+            >>> params = model.to_bf16(model.params, mask)
```
```suggestion
            >>> model.params = model.to_bf16(model.params, mask)
```

Same as above (assuming I understood correctly that `to_bf16(...)` is **not** an in-place operation

patrickvonplaten commented on src/transformers/modeling_flax_utils.py:
```diff
@@ -154,6 +154,115 @@ def params(self, params: Union[Dict, FrozenDict]):
+            >>> # we'll first cast to fp16 and back to fp32
+            >>> params = model.to_f16(model.params)
+            >>> # now cast back to fp32
+            >>> params = model.to_fp32(params)
```
same comments as above -> I'd maybe directly set it to `model.params` 

Also let's make sure that this is shown in the docs

patrickvonplaten commented on src/transformers/modeling_flax_utils.py:
```diff
@@ -154,6 +154,115 @@ def params(self, params: Union[Dict, FrozenDict]):
+    def to_fp32(self, params: Union[Dict, FrozenDict], mask: Any = None):
+        r"""
+        Cast the floating-point ``parmas`` to ``jax.numpy.float32``. This method can be used to explicitly convert the
+        model paramters to ``float32``.
```
```suggestion
        model parameters to fp32 precision
```

patrickvonplaten commented on src/transformers/modeling_flax_utils.py:
```diff
@@ -154,6 +154,115 @@ def params(self, params: Union[Dict, FrozenDict]):
+    def to_fp16(self, params: Union[Dict, FrozenDict], mask: Any = None):
+        r"""
+        Cast the floating-point ``parmas`` to ``jax.numpy.float16``. This method can be used to explicitly convert the
+        model paramters to ``float16``.
```
```suggestion
        model parameters to float16 precision.
```

patrickvonplaten reviewed (approved):
Great, the functionality looks good. Left a couple of suggestions for improvement for the docstring. I wonder whether we should even leave a couple more sentences under `to_bfloat16(...)` and `to_float16(...)` that state **why** one would call such a method. E.g. bfloat16 for TPU to save memory and float16 for GPU to save memory. Should we maybe also adapt the docstring of the `dtype` input that one can pass to a Flax model to really make sure that people understand that `dtype` **never** influences the weights but just the computations?

Also would be great to check that the docstring is nicely shown in the docs .

Finally, we should leave a big statement in the PR description with :rotating_light: that this PR is backwards breaking.

We should really bring across the message that `dtype` == computation precision & `to_...()` converts parameters

patrickvonplaten commented:
Thanks for finishing the PR!

patil-suraj committed:
Apply suggestions from code review

Co-authored-by: Patrick von Platen <patrick.v.platen@gmail.com>

patil-suraj committed:
fix doc

patil-suraj committed:
fix doc

patil-suraj committed:
add detailed docstring for dtype parameter

patrickvonplaten commented:
Think we just need to update the Flax templates now and we're good to go :-)

patil-suraj commented:
The flax templates are updated, not sure why the tests are still failing.

patrickvonplaten commented on templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_flax_{{cookiecutter.lowercase_modelname}}.py:
```diff
@@ -17,6 +17,7 @@
 {% if cookiecutter.is_encoder_decoder_model == "False" %}
 
 from typing import Callable, Optional, Tuple
+from jax._src.dtypes import dtype
```
Think this is not needed no?

patrickvonplaten replied:
It wasn't required before either

patil-suraj replied:
```suggestion
```

patrickvonplaten commented on src/transformers/modeling_flax_utils.py:
```diff
@@ -184,6 +300,19 @@ def from_pretrained(
+                This can be used to enable mixed-precision training or half-precision inference on GPUs or TPUs. If
+                specified all the computation will be performed with the given ``dtype``.
+
+                **Note that this only specifies the dtype of the computation and does not influence the dtype of model
```
(nit)
```suggestion
                **Note that this only specifies the precision at which the computation is done and not the precision of the parameter weights
```

patil-suraj replied:
flax uses the `precision` parameter to indicate numerical precision which is different from `dtype`, 
see https://github.com/google/flax/blob/a9bdc363ca9aee5051291e9bc421c6de7b2b5dbc/flax/linen/linear.py#L62
so I would stick to `dtype` here.

patrickvonplaten commented on src/transformers/modeling_flax_utils.py:
```diff
@@ -184,6 +300,19 @@ def from_pretrained(
                       :func:`~transformers.FlaxPreTrainedModel.save_pretrained`, e.g., ``./my_model_directory/``.
                     - A path or url to a `pt index checkpoint file` (e.g, ``./tf_model/model.ckpt.index``). In this
                       case, ``from_pt`` should be set to :obj:`True`.
+            dtype (:obj:`jax.numpy.dtype`, `optional`, defaults to :obj:`jax.numpy.float32`):
```
Very Nice!

patil-suraj committed:
remove un-necessary import
